### PR TITLE
Fix pickle/deepcopy not preserve original edge indices

### LINKED ISSUE #585: Serialize and deserialize graphs change order of the edges
### What is the expected enhancement?

The current serialization mechanism of PyDiGraph doesn't keep the original order of edges in the graph. This creates an issue for libraries that store the edge index inside the edge weight object (why need to store the edge index inside the edge weight object? it is easier to work with for graphs in which nodes/edges have lots of properties such as Wikidata). A graph is serialized and deserialized when using it in a multiprocess environment. After deserializing, the order of the edges changes and so edge indices stored inside edge weights do not match the new indices, thus messing up the graph.

I took a look at the serialization code here: https://github.com/Qiskit/retworkx/blob/a0aaa1f93ef21fe77dccee63cd37fdb8f43c7cee/src/digraph.rs#L283-L291 and I think we can maintain the original order by dumping the edges in a separated loop. This won't increase the runtime and maybe faster.

If you are okay with my propose, I am happy to create a PR for this if you don't have time. Thanks.

### PR DESCRIPTION
**Pull request by binh-vu**
This PR fixes issue #585. Now pickling graph & digraph should keep the original edge index.

Did a simple benchmark with `directed_mesh_graph(300)` (no node removed and 50 nodes removed -- 1/6 of the graph). Performance is very similar to previous implementation. If we do not need to maintain the original edge index, the new code is ~10% faster. The serialized data size of the graph is almost the same as previous implementation if no edge is removed. When 1/6 of the graph is removed, the serialized size is increased by about 6%.

Fixes #585

### CODE REVIEW

**Review by mtreinish — COMMENTED** [review]
Thanks for pushing this fix, overall it LGTM. I have a few inline comments most are just idle thoughts or questions. I think the only real blocker is maintaining support with the old pickle format.

Also while you're updating the branch would you mind adding a fix release note. You can see the process for this here: https://github.com/Qiskit/retworkx/blob/main/CONTRIBUTING.md#release-notes

File: src/digraph.rs
Comment (binh-vu) [reply src/digraph.rs:305]: I actually don't know exactly how `node_removed` is implemented, but I saw the [original implementation](https://github.com/Qiskit/retworkx/blob/add8e5b8ad9026551acd2ac7eda1bd955ca656c1/src/digraph.rs#L323) also reset `node_removed` so I assume that there is no guarantee that `node_removed` is always correct and needed recalculation (e.g., delete n node and add back n node, `node_removed` is not set to false).

File: src/digraph.rs
Comment (binh-vu) [reply src/digraph.rs:295]: My original implementation is actually similar to what you said, but I saw a significant increase so I switch to this. Unless you see >50% edges are removed, the *minimum* size increase is (pointer size) * #edges > (null size) * #removed_edges.

Also, IMO doing iteration over a list will be faster than doing it over the range and we don't have to pay for extra cost for dictionary access and reconstruct the dictionary at the beginning.

File: src/digraph.rs
Comment (binh-vu) [reply src/digraph.rs:314]: If backward compatibility is a *must* (and there are users needing it), then I'm ok. But IMO, if they serialize the graph and store it for later use, we should have a separated function for writing graph like networkx. Relying on pickle is not a good idea because the internal properties of a class may change over time (rename, remove, or add property) and maintaining different versions is just lots of work

File: src/digraph.rs
Comment (binh-vu) [reply src/digraph.rs:437]: It's correct. Let's me fix it.

File: src/digraph.rs
Comment (binh-vu) [reply src/digraph.rs:397]: I don't like that too. What I was thinking is to create inline functions for `get_index`, and `get_value`. But I don't know where to put it. May be at the end of the file?

File: src/digraph.rs
Comment (mtreinish) [reply src/digraph.rs:305]: It is enforced at retworkx's API. So if things are working on petgraph objects directly the attribute isn't updated for us. But if you use the methods on the pyclass it would. So for `__setstate__` we had to set this manually because we create a petgraph graph directly. But from the python api perspective we should always be tracking this correctly (if not it's a bug).

File: src/digraph.rs
Comment (mtreinish) [reply src/digraph.rs:314]: I generally agree with this, but it wouldn't be too hard to do in this case. But either way we can hold off on it for this PR I can look at this separately.

**Comment by dizhouwu** [comment]
Any chance this would get merged?

**Comment by mtreinish** [comment]
> Any chance this would get merged?

There were some issues I outlined in review that I thought needed to be updated before we merged. But it looks like the has stalled since my initial review. I'll take a quick look to see if I can rebase it and then fix the issues I outlined.

File: src/digraph.rs
Comment (mtreinish) [reply src/digraph.rs]: Done in: https://github.com/Qiskit/rustworkx/pull/589/commits/73e6594136e74f16274f258e9f799c5be4c3477a

**Review by mtreinish — APPROVED** [review]

**Comment by dizhouwu** [comment]
Thank you so much for merging this - wondering if we could be getting a release soon?

**Pull request merged**
